File: Design-Patterns-Using-Ruby/Flyweight/Flyweight2.rb
# Flyweight interface
class Flyweight
    def operation(extrinsic_state)
        raise NotImplementedError, 'Subclasses must implement the operation method'
    end
end

# Concrete Flyweight class
class ConcreteFlyweight < Flyweight
    attr_reader :intrinsic_state

    def initialize(intrinsic_state)
        @intrinsic_state = intrinsic_state
    end

    def operation(extrinsic_state)
        puts "Operation inside concrete flyweight: #{self}"
    end
end

# FlyweightFactory class
class FlyweightFactory
    def initialize
        @flyweights = {}
    end

    def get_flyweight(intrinsic_state)
        @flyweights[intrinsic_state] ||= ConcreteFlyweight.new(intrinsic_state)
    end
end

# ClientClass
class ClientClass
    attr_reader :flyweight, :extrinsic_state

    def initialize(factory, intrinsic_state, extrinsic_state)
        @flyweight = factory.get_flyweight(intrinsic_state)
        @extrinsic_state = extrinsic_state
    end

    def operation
        puts "Operation inside context: #{self}"
        flyweight.operation(extrinsic_state)
    end
end

# Client code
factory = FlyweightFactory.new
c = ClientClass.new(factory, 'common', 'separate1')
c.operation
c2 = ClientClass.new(factory, 'common', 'separate2')
c2.operation

=begin 
Operation inside context: #<ClientClass:0x00007f216104bcb0>
Operation inside concrete flyweight: #<ConcreteFlyweight:0x00007f216104b440>
Operation inside context: #<ClientClass:0x00007f2161049050>
Operation inside concrete flyweight: #<ConcreteFlyweight:0x00007f216104b440>
=end

File: Design-Patterns-Using-Swift/Flyweight/Flyweight2.swift
// Flyweight protocol
protocol Flyweight {
    func operation(extrinsicState: String)
}

// Concrete Flyweight class
class ConcreteFlyweight: Flyweight {
    private let intrinsicState: String

    init(intrinsicState: String) {
        self.intrinsicState = intrinsicState
    }

    func operation(extrinsicState: String) {
        print("Operation inside concrete flyweight: \(self)")
    }
}

// FlyweightFactory class
class FlyweightFactory {
    private var flyweights: [String: Flyweight] = [:]

    func getFlyweight(intrinsicState: String) -> Flyweight {
        if let existingFlyweight = flyweights[intrinsicState] {
            return existingFlyweight
        } else {
            let newFlyweight = ConcreteFlyweight(intrinsicState: intrinsicState)
            flyweights[intrinsicState] = newFlyweight
            return newFlyweight
        }
    }
}

// ClientClass
class ClientClass {
    var flyweight: Flyweight
    var extrinsicState: String

    init(factory: FlyweightFactory, intrinsicState: String, extrinsicState: String) {
        self.flyweight = factory.getFlyweight(intrinsicState: intrinsicState)
        self.extrinsicState = extrinsicState
    }

    func operation() {
        print("Operation inside context: \(self)")
        flyweight.operation(extrinsicState: extrinsicState)
    }
}

// Client Code
let factory = FlyweightFactory()
let c = ClientClass(factory: factory, intrinsicState: "common", extrinsicState: "separate1")
c.operation()
let c2 = ClientClass(factory: factory, intrinsicState: "common", extrinsicState: "separate2")
c2.operation()

/*
Operation inside context: Flyweight2.ClientClass
Operation inside concrete flyweight: Flyweight2.ConcreteFlyweight
Operation inside context: Flyweight2.ClientClass
Operation inside concrete flyweight: Flyweight2.ConcreteFlyweight
*/